go/merklestore/store.go

```diff
@@ -0,0 +1,346 @@
+		// The root is still too old, even after an attempted refresh.
+		m.CDebugf("MerkleStore: merkle root too old")
+		return "", "", NewMerkleStoreError("merkle root too old: %v %s", seqnoWrap(root.Seqno()), root.Fetched())
+	}
+
+	// This is the hash we are being instructed to use.
+	hash := keybase1.MerkleStoreKitHash(s.getRootHash(*root))
+
+	if hash == "" {
+		return "", "", NewMerkleStoreError("merkle root has empty %s hash: %v", s.tag, seqnoWrap(root.Seqno()))
+	}
+
+	// Use in-memory cache if it matches
+	if fromMem := s.memGet(hash); fromMem != nil {
+		m.CDebugf("MerkleStore: mem cache hit, using hash: %s", hash)
+		return *fromMem, hash, nil
+	}
+
+	// Use db cache if it matches
+	if fromDB := s.dbGet(m, hash); fromDB != nil {
+		m.CDebugf("MerkleStore: db cache hit")
+
+		// Store to memory
+		s.memSet(hash, *fromDB)
+
+		m.CDebugf("MerkleStore: using hash: %s", hash)
+		return *fromDB, hash, nil
+	}
+
+	// Fetch from the server
+	// This validates the hash
+	kitJSON, err := s.fetch(m, hash)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Store to memory
+	s.memSet(hash, kitJSON)
+
+	// db write
+	s.dbSet(m.BackgroundWithLogTags(), hash, kitJSON)
+
+	m.CDebugf("MerkleStore: using hash: %s", hash)
+	return kitJSON, hash, nil
+}
+
+type merkleStoreServerRes struct {
+	Status  libkb.AppStatus         `json:"status"`
+	KitJSON keybase1.MerkleStoreKit `json:"kit_json"`
+}
+
+func (r *merkleStoreServerRes) GetAppStatus() *libkb.AppStatus {
+	return &r.Status
+}
+
+// Fetch data and check the hash.
+func (s *MerkleStoreImpl) fetch(m libkb.MetaContext, hash keybase1.MerkleStoreKitHash) (keybase1.MerkleStoreKit, error) {
+	m.CDebugf("MerkleStore: fetching from server: %s", hash)
+	var res merkleStoreServerRes
+	err := m.G().API.GetDecode(libkb.APIArg{
+		Endpoint:    s.endpoint,
+		SessionType: libkb.APISessionTypeNONE,
+		MetaContext: m,
+		Args: libkb.HTTPArgs{
+			"hash": libkb.S{Val: string(hash)},
+		},
+	}, &res)
+	if err != nil {
+		return "", NewMerkleStoreError(err.Error())
+	}
+	if res.KitJSON == "" {
+		return "", NewMerkleStoreError("server returned empty kit for %s", s.tag)
+	}
+	if s.hash(res.KitJSON) != hash {
+		m.CDebugf("%s hash mismatch: got:%s expected:%s", s.tag, s.hash(res.KitJSON), hash)
+		return "", NewMerkleStoreError("server returned wrong kit for %s", s.tag)
+	}
+	return res.KitJSON, nil
+}
+
+// updateRoot kicks MerkleClient to update its merkle root
+// by doing a LookupUser on some arbitrary user.
+func (s *MerkleStoreImpl) refreshRoot(m libkb.MetaContext) error {
+	q := libkb.NewHTTPArgs()
+	// The user lookup here is unnecessary. It is done because that is what is
+	// easy with MerkleClient.  The user looked up is you if known, otherwise
+	// arbitrarily t_alice.  If t_alice is removed, this path will break.
+	uid := s.G().GetMyUID()
+	if len(uid) == 0 {
+		// Use t_alice's uid.
+		uid = libkb.TAliceUID
+	}
+	q.Add("uid", libkb.UIDArg(uid))
+	_, err := s.G().MerkleClient.LookupUser(m, q, nil)
+	return err
+}
```

Comment:
yeck. Maybe there's a new function on MerkleClient to accomplish the same thing? @maxtaco

Reply:
yeah this is unfortunate. if there is something existing that is better we can swap it in or open a ticket to do the same but won't block this from merging